core/unit_test/view/TestViewCustomizationDataHandle.hpp

```diff
@@ -0,0 +1,232 @@
+                !Kokkos::Impl::IsReferenceCountedAccessor<DstAccessor>::value &&
+                std::is_convertible_v<NestedAccessor, DstAccessor>>>
+  KOKKOS_FUNCTION operator DstAccessor() const {
+    return m_nested_acc;
+  }
+
+  KOKKOS_FUNCTION
+  constexpr reference access(
+#ifndef KOKKOS_ENABLE_OPENACC
+      const data_handle_type& p,
+#else
+      // FIXME OpenACC: illegal address when passing by reference
+      data_handle_type p,
+#endif
+      size_t i) const {
+    return m_nested_acc.access(p.get(), i);
+  }
+
+  KOKKOS_FUNCTION
+  constexpr data_handle_type offset(
+#ifndef KOKKOS_ENABLE_OPENACC
+      const data_handle_type& p,
+#else
+      // FIXME OpenACC: illegal address when passing by reference
+      data_handle_type p,
+#endif
+      size_t i) const {
+    return data_handle_type{p, m_nested_acc.offset(p.get(), i)};
+  }
+
+  KOKKOS_FUNCTION
+  constexpr auto nested_accessor() const { return m_nested_acc; }
+
+ private:
+#ifdef MDSPAN_IMPL_NO_UNIQUE_ADDRESS
+  MDSPAN_IMPL_NO_UNIQUE_ADDRESS
+#else
+  [[no_unique_address]]
+#endif
+  NestedAccessor m_nested_acc;
+};
+
+template <typename DataHandleType, typename MDSpanExtents,
+          typename MDSpanLayoutType, typename ElementType, typename MemorySpace>
+KOKKOS_INLINE_FUNCTION constexpr DataHandleType data_handle_from_allocation(
+    Kokkos::Impl::SharedAllocationRecord<void, void>* rec,
+    const typename MDSpanLayoutType::template mapping<
+        MDSpanExtents>& /* mapping */,
+    const Kokkos::Impl::SpaceAwareAccessor<
+        MemorySpace,
+        TestCustomDataHandleAccessor<
+            ElementType, MDSpanExtents, MDSpanLayoutType, MemorySpace,
+            Kokkos::default_accessor<ElementType>>>& /* accessor */) {
+  return DataHandleType(rec, 13);
+}
+}  // namespace TestDataHandle
+
+// Don't try this at home!
+namespace Kokkos::Impl {
+template <class Traits, class LayoutType>
+struct MDSpanViewTraits<Traits, TestDataHandle::CustomDataHandleTag_t,
+                        LayoutType> {
+  using index_type =
+      typename MDSpanViewTraits<Traits, void, LayoutType>::index_type;
+  using extents_type =
+      typename MDSpanViewTraits<Traits, void, LayoutType>::extents_type;
+  using mdspan_layout_type =
+      typename MDSpanViewTraits<Traits, void, LayoutType>::mdspan_layout_type;
+  using accessor_type = Kokkos::Impl::SpaceAwareAccessor<
+      typename Traits::memory_space,
+      TestDataHandle::TestCustomDataHandleAccessor<
+          typename Traits::value_type, extents_type, mdspan_layout_type,
+          typename Traits::memory_space,
+          Kokkos::default_accessor<typename Traits::value_type>>>;
+  // This will static assert that accessor_type is legal
+  using mdspan_type = mdspan<typename Traits::value_type, extents_type,
+                             mdspan_layout_type, accessor_type>;
+};
+
+template <class ElementType, class MemorySpace>
+KOKKOS_INLINE_FUNCTION constexpr auto ptr_from_data_handle(
```

Comment:
Are we going to write this down somewhere? I mean we are ADL-ing stuff, injecting traits and overloading functions in `Impl` to make this work ... And it only does if all of it is there ...

Reply:
oh this is actually not ADL'd, `data_handle_from_allocation` is